pink/barriers/barrier.py

```diff
@@ -0,0 +1,221 @@
+        """
+        return np.zeros(configuration.model.nv)
+
+    def compute_qp_objective(
+        self,
+        configuration: Configuration,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        r"""Compute the quadratic objective function for the barrier-based QP.
+
+        The quadratic objective function includes a regularization term
+        based on the safe backup policy:
+
+        .. math::
+
+            \gamma(\boldsymbol{q})\left\| \dot{\boldsymbol{q}}-
+            \dot{\boldsymbol{q}}_{safe}(\boldsymbol{q})\right\|^{2}
+
+        where :math:`\gamma(\boldsymbol{q})` is a configuration-dependent
+        weight and :math:`\dot{\boldsymbol{q}}_{safe}(\boldsymbol{q})`
+        is the safe backup policy.
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+            dt: Time step for discrete-time implementation. Defaults to 1e-3.
+
+        Returns:
+            Tuple containing the quadratic objective matrix (H) and linear
+                objective vector (c).
+        """
+        jac = self.compute_jacobian(configuration)
+        H = np.zeros((configuration.model.nv, configuration.model.nv))
+        # c = 1e-3 * np.linalg.norm(jac, axis=0) / dt
+        c = np.zeros(configuration.model.nv)
+
+        if self.r > 1e-6:
+            self.safe_policy = self.compute_safe_policy(configuration)
+
+            H += (
+                self.r
+                / (np.linalg.norm(jac) ** 2)
+                * np.eye(configuration.model.nv)
+            )
+            c += -2 * self.r / (np.linalg.norm(jac) ** 2) * self.safe_policy
+
+        return (H, c)
+
+    def compute_qp_inequality(
+        self,
+        configuration: Configuration,
+        dt: float = 1e-3,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        r"""Compute the linear inequality constraints for the barrier-based QP.
+
+        The linear inequality constraints enforce the barrier conditions:
+
+        .. math::
+
+            \frac{\partial \boldsymbol{h}_j}
+            {\partial \boldsymbol{q}} \dot{\boldsymbol{q}} +
+            \alpha_j(\boldsymbol{h}_j(\boldsymbol{q})) \geq 0, \quad \forall j
+
+        where :math:`\frac{\partial \boldsymbol{h}_j}{\partial \boldsymbol{q}}`
+        are the Jacobians of the constraint functions,
+        :math:`\dot{\boldsymbol{q}}` is the joint velocity vector,
+        and :math:`\alpha_j` are extended class K functions.
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+            dt: Time step for discrete-time implementation. Defaults to 1e-3.
+
+        Returns:
+            Tuple containing the inequality constraint matrix (G)
+                and vector (h).
+        """
+        G = -self.compute_jacobian(configuration) / dt
+        barrier_value = self.compute_barrier(configuration)
+        h = np.array(
+            [
+                self.gain[i] * self.class_k_fn(barrier_value[i])
```

Comment:
It looks like the gain is redundant with the class K function here, as the function is general enough to include the gain inside it. Maybe you found it convenient to keep a gain as well for easier tuning?

Reply:
Yes, this is definitely has proven to be more convenient. Moreover, practice've shown that it is convenient to set specific class kappa function for specific implementation of barrier (for example, joint barriers, that are yet to come in the following PRs) always should use function:
$$f(h(q)) = \frac{h(q)}{1 + |h(q)|}$$
to reduce effect of barrier near the limit (for instance, to reduce chattering).

However, sometimes it is more convenient to leave the possibility to construct and set your own class kappa function at barrier initialisation, that's why it is an argument.